Vectorize ProbabilisticMap.IndexOfAny

MihaZupan opened the pull request:
#### AVX2

|             Method | Toolchain | NewLineFrequency | Length |       Mean |     Error | Ratio |
|------------------- |---------- |----------------- |------- |-----------:|----------:|------:|
| ReplaceLineEndings |      main |                0 |  10000 |  11.481 us | 0.0569 us |  1.00 |
| ReplaceLineEndings |        pr |                0 |  10000 |   1.086 us | 0.0026 us |  0.09 |
|                    |           |                  |        |            |           |       |
| ReplaceLineEndings |      main |             0.05 |  10000 |  24.280 us | 0.0478 us |  1.00 |
| ReplaceLineEndings |        pr |             0.05 |  10000 |  13.983 us | 0.0596 us |  0.58 |
|                    |           |                  |        |            |           |       |
| ReplaceLineEndings |      main |              0.1 |  10000 |  35.616 us | 0.1959 us |  1.00 |
| ReplaceLineEndings |        pr |              0.1 |  10000 |  27.322 us | 0.3056 us |  0.77 |
|                    |           |                  |        |            |           |       |
| ReplaceLineEndings |      main |              0.2 |  10000 |  58.372 us | 1.0346 us |  1.00 |
| ReplaceLineEndings |        pr |              0.2 |  10000 |  52.060 us | 0.2331 us |  0.89 |
|                    |           |                  |        |            |           |       |
| ReplaceLineEndings |      main |                1 |  10000 | 160.760 us | 1.1764 us |  1.00 |
| ReplaceLineEndings |        pr |                1 |  10000 | 201.541 us | 4.8189 us |  1.26 |

#### ARM64

| Method | Length |      Mean |     Error |
|------- |------- |----------:|----------:|
|    Old |  10000 | 13.122 µs | 0.0117 µs |
|    New |  10000 |  3.701 µs | 0.0012 µs |

<details>
<summary>Benchmark source</summary>

```c#
public class ReplaceLineEndingsBenchmark
{
    private string _input;

    [Params(0.0, 0.05, 0.1, 0.2, 1.0)]
    public double NewLineFrequency;

    [Params(10_000)]
    public int Length;

    [GlobalSetup]
    public void Setup()
    {
        char[] input = new char[Length];

        var rng = new Random(42);

        for (int i = 0; i < input.Length; i++)
        {
            if (rng.NextDouble() < NewLineFrequency)
            {
                input[i] = '\n';
            }
            else
            {
                char c;
                do
                {
                    c = (char)rng.Next(0, 65536);
                }
                while ("\r\n\f\u0085\u2028\u2029".Contains(c));

                input[i] = c;
            }
        }

        _input = new string(input);
    }

    [Benchmark]
    public string ReplaceLineEndings() => _input.ReplaceLineEndings();
}
```

</details>

If we care, we could also do this for `LastIndexOfAny`.

#80297 could be interesting for ARM if we could do a 256-bit lookup instead of blending together smaller lookups.

MihaZupan committed:
Vectorize ProbabilisticMap.IndexOfAny on AVX2

MihaZupan committed:
Use ResetLowestSetBit from BitOperations

MihaZupan committed:
Speed up Avx2 and add Vector128 support

MihaZupan commented:
/azp run runtime-libraries-coreclr outerloop

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +70,123 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector256<byte> ContainsMask32CharsAvx2(Vector256<byte> charMapLower, Vector256<byte> charMapUpper, ref char searchSpace)
+        {
+            Vector256<ushort> source0 = Vector256.LoadUnsafe(ref Unsafe.As<char, ushort>(ref searchSpace));
```
Not the first time I see this pattern - I think it's time to introduce an internal (for now) LoadUnsafe overload for `char`

tannergooding replied:
`LoadUnsafe` that takes a `char` and returns `ushort` probably won't pass API review
I previously suggested we just add `char` support directly and it was shot down (CC. @GrabYourPitchforks)

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +70,123 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<byte> Shuffle(Vector128<byte> vector, Vector128<byte> indices)
```
I already see 3 helpers in BCL for this - can we unite them? e.g. internal API in Vector128

MihaZupan replied:
`Vector128.NativeShuffle`? Or maybe platform?

Or would it make sense to create a different place to dump internal utils? Another example that comes to mind is the `Avx2.Permute4x64` helper that I've added for the 3rd time now.

EgorBo replied:
cc @tannergooding

tannergooding replied:
`ShuffleUnsafe` might be a better name and document it as not having deterministic behavior for out of range indices might be better (we've exclusively used `Unsafe` for non-deterministic or potentially dangerous behavior so far, `Native` is typically reserved for other things)

I'd be willing to champion it if someone opened a proposal

MihaZupan committed:
Add Vector{128/256}.LoadUnsafe(ref char) and Vector128.ShuffleUnsafe

MihaZupan commented:
Added `Vector{128/256}.LoadUnsafe(ref char)` and `Vector128.ShuffleUnsafe` as internal APIs for now and replaced a few existing uses. I intentionally excluded the shuffle helper in [`Base64Decoder`](https://github.com/dotnet/runtime/blob/8ba2ad4f46e8011331d29dacf416a8647caa0bbb/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L482-L494) as it feels more readable as-is.

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/Guid.cs:
```diff
@@ -1219,8 +1219,8 @@ public bool TryFormat(Span<char> destination, out int charsWritten, [StringSynta
                         Vector128<byte> nibbles = Vector128.ShiftRightLogical(srcVec.AsUInt64(), 4).AsByte();
                         Vector128<byte> lowNibbles = UnpackLow(nibbles, srcVec) & Vector128.Create((byte)0xF);
                         Vector128<byte> highNibbles = UnpackHigh(nibbles, srcVec) & Vector128.Create((byte)0xF);
-                        (Vector128<ushort> v0, Vector128<ushort> v1) = Vector128.Widen(Shuffle(hexMap, lowNibbles));
```
Can you revert changes in this file (Guid.cs) since I overwrote them in https://github.com/dotnet/runtime/pull/81666 so we can avoid a conflict

MihaZupan replied:
Reverted

MihaZupan committed:
Use Vector128.ShuffleUnsafe in more places

lewing commented:
@radekdoulik we should add PackedSimd versions of the Unsafe* functions wasm supports to ease handling all the calling callers.  I'm not sure we have a great pattern to simplify the *.IsHardwareAccelerated paths that works for all the cases yet but that is worth considering too.

MihaZupan commented:
Any thoughts on this @dotnet/area-system-memory?

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs:
```diff
@@ -477,20 +477,17 @@ private static unsafe void Avx2Decode(ref byte* srcBytes, ref byte* destBytes, b
-            else
-            {
-                return AdvSimd.Arm64.VectorTableLookup(left, Vector128.BitwiseAnd(right, mask8F));
+                right &= mask8F;
```
honestly speaking I'm not sure this operation is even needed here for base64 but it's irrelevant to this PR

MihaZupan replied:
cc: @gfoidl would you mind checking if that's the case? Looks like the indices are already being masked by the callers in a few ways, but I am not familiar enough with the base64 logic here.
If it is redundant, we can remove the `SimdShuffle` helper entirely.

EgorBo replied:
I bet it was added during porting that algorithm to ARM64 and to adjust `VectorTableLookup` to behave like SSSE3 Shuffle in a general case

gfoidl replied:
If ARM for the `VectorTableLookup` only takes the low nibble into account (I don't know) it's not needed, and after https://github.com/dotnet/runtime/pull/76795 there should be enough tests to cover non-ASCII input cases well.
In the code the masking of the input-nibbles happens with `0x2F` https://github.com/dotnet/runtime/blob/c2b12d7fce16b9e4ec3a44bf747118732e85782c/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L591-L592 instead of `0xF`, as that constant already is around thus saving the use of another register. Then after the check for invalid input, only valid values are around, so further down the code no bits in the high-nibble exist.

If `VectorTableLookup` takes the higher nibble into account, then masking with `0xF` instead of `0x8F` is fine.

> behave like SSSE3 Shuffle in a general case

This seems very plausible, as `0x8F` has the highest bit set with has special meaning in SSSE3 shuffle.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -23,8 +26,19 @@ namespace System.Buffers
-        private const int IndexMask = 0x7;
-        private const int IndexShift = 0x3;
+        // We need at least Sse41 for the hardware accelerated ConditionalSelect support.
+        private static bool IsVectorizationSupported => Sse41.IsSupported || AdvSimd.Arm64.IsSupported;
```
This is not `readonly` which means it incurs an actual check per use.

Making it a `static readonly` is also strictly worse sicne these are otherwise treated as JIT time constants (always, even for debug code).

tannergooding replied:
Ah, I misread. It's a property.

Still believe you should keep these as direct usages so the normal benefits of these being treated as constants kicks in.

stephentoub replied:
> This is not readonly which means it incurs an actual check per use.

This won't get inlined and be equivalent to if you wrote `if (Sse41.IsSupported || AdvSimd.Arm64.IsSupported)` at the call site?

stephentoub replied:
Nevermind, my question overlapped with your revision.

gfoidl replied:
It's a property and not a field, so should be treated as JIT constant.

MihaZupan replied:
I can move the check inline, though it's a bit unfortunate to have to duplicate it.

Btw I've used the same pattern (property) in IndexOfAnyAsciiSearcher and PackedSpanHelpers already. Do you think those should also be updated? It'd be great if the JIT could just treat these as equivalent so that wasn't necessary ofc.

tannergooding replied:
> It'd be great if the JIT could just treat these as equivalent so that wasn't necessary ofc.

Right, the general issue is in the cost required to resolve non-intrinsic methods, scan the IL (which then requires further resolution of methods - this time intrinsic), and then make a decision around whether inlining is profitable or not.

There are many factors that can make this fail or be less than optimal and its a balance between what's good for typical user code and what's good for highly optimized hot code.

> Btw I've used the same pattern (property) in IndexOfAnyAsciiSearcher and PackedSpanHelpers already

There is likely some benefit to doing so, long term. It's also likely not the "worst thing" for these methods and the JIT is probably doing the "right thing" in Tier 1. It's just having to do more to make it happen

tannergooding replied:
In the case of intrinsic methods, the VM flags them specially and so resolution tends to be a lot cheaper and we can make an easy determination of "this is more than likely profitable".

We can also avoid the inliner entirely by creating new IR nodes directly. In the case of `Isa.IsSupported`, this always produces a constant `true` or constant `false` node and so its as if the user had directly written `if (true)` or `if (false)` themselves.

MihaZupan replied:
Moved the checks inline in this file

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +70,114 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+                (source1 & Vector256.Create((ushort)255)).AsInt16());
+
+            Vector256<byte> sourceUpper = Avx2.PackUnsignedSaturate(
+                Vector256.ShiftRightLogical(source0, 8).AsInt16(),
+                Vector256.ShiftRightLogical(source1, 8).AsInt16());
```
You could use the operators here:
```suggestion
                (source0 >>> 8).AsInt16(),
                (source1 >>> 8).AsInt16());
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +70,114 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector256<byte> IsCharBitNotSetAvx2(Vector256<byte> charMapLower, Vector256<byte> charMapUpper, Vector256<byte> values)
+        {
+            Vector256<byte> highNibble = Avx2.ShiftRightLogical(values.AsInt32(), VectorizedIndexShift).AsByte() & Vector256.Create((byte)15);
```
Same here, could be:
```suggestion
            Vector256<byte> highNibble = (values.AsInt32() >>> VectorizedIndexShift).AsByte() & Vector256.Create((byte)15);
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +70,114 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+                ? Sse2.PackUnsignedSaturate((source0 & Vector128.Create((ushort)255)).AsInt16(), (source1 & Vector128.Create((ushort)255)).AsInt16())
+                : AdvSimd.Arm64.UnzipEven(source0.AsByte(), source1.AsByte());
+
+            Vector128<byte> sourceUpper = Sse2.IsSupported
+                ? Sse2.PackUnsignedSaturate(Vector128.ShiftRightLogical(source0, 8).AsInt16(), Vector128.ShiftRightLogical(source1, 8).AsInt16())
```
and here:
```suggestion
            Vector128<byte> sourceUpper = Sse2.IsSupported
                ? Sse2.PackUnsignedSaturate((source0 >>> 8).AsInt16(), (source1 >>> 8).AsInt16())
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +70,114 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<byte> IsCharBitNotSet(Vector128<byte> charMapLower, Vector128<byte> charMapUpper, Vector128<byte> values)
+        {
+            Vector128<byte> highNibble = Sse2.IsSupported
+                ? Sse2.ShiftRightLogical(values.AsInt32(), VectorizedIndexShift).AsByte() & Vector128.Create((byte)15)
+                : AdvSimd.ShiftRightLogical(values, VectorizedIndexShift);
```
These are different because shifting via `byte` isn't accelerated on xarch, right?

MihaZupan replied:
Yes. We do the same thing in IndexOfAnyAsciiSearcher as well, where that also feeds into a shuffle

tannergooding replied:
👍, probably worth opening an issue so an efficient implementation can be done and we can avoid the separate paths here.

MihaZupan replied:
https://github.com/dotnet/runtime/issues/82564

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -201,15 +306,115 @@ internal static int IndexOfAny<TNegator>(ref uint charMap, ref char searchSpace,
             ref char cur = ref searchSpace;
 
-            while (!Unsafe.AreSame(ref cur, ref searchSpaceEnd))
+            if (IsVectorizationSupported && typeof(TNegator) == typeof(IndexOfAnyAsciiSearcher.DontNegate) && searchSpaceLength >= 16)
```
This is a pretty big body to be aggressively inlined and may actually hurt inlining for the other path.

Should this be moved to separate helper method which has its own inlining, that way it can be more easily dropped as dead code when `TNegator` isn't `DontNegate`?

MihaZupan replied:
I can move it out, but I plan on reworking how this logic is set up soon anyway.
The fact we're using ProbabilisticMap for the -Except case was a mistake, so this will end up being just one path, without the generic negator.

MihaZupan committed:
PR feedback

MihaZupan committed:
Replace another ShiftRightLogical with '>>>'

MihaZupan commented:
Failures are: #82575, #81123, #82611

MihaZupan commented:
Anything else that should be changed here, or is this one good to merge @tannergooding?

MihaZupan commented:
@tannergooding can this be merged? (I'm trying to avoid more merge conflicts as this was already rebased a fair number of times)

MihaZupan committed:
Add WASM path to Vector128.ShuffleUnsafe

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/ProbabilisticMap.cs:
```diff
@@ -56,23 +69,116 @@ public ProbabilisticMap(ReadOnlySpan<char> values)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector256<byte> IsCharBitNotSetAvx2(Vector256<byte> charMapLower, Vector256<byte> charMapUpper, Vector256<byte> values)
```
Nit: Can we avoid the double-negative?

MihaZupan replied:
Doing so is slightly more expensive (we're talking about ~2% throughput), so using the more readable version may make sense.

Theoretically, a sufficiently smart optimizer could get that difference back.
Essentially it's the difference between `~(A | B)` vs `(~A) & (~B)` where all elements of A and B are known to be either 0 or FF.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs:
```diff
@@ -2419,6 +2435,32 @@ public static Vector128<sbyte> Shuffle(Vector128<sbyte> vector, Vector128<sbyte>
         }
 
+        /// <summary>Creates a new vector by selecting values from an input vector using a set of indices.
+        /// Behavior is platform-dependent for out-of-range indices.</summary>
```
Right. For the xplat hardware intrinsics we strive to expose deterministic functionality as otherwise it easily leads to bugs and other subtleties between platforms.

We expose `Unsafe` variants to allow users to"opt-in" to non-deterministic functionality, such as leaving upper-bits undefined or doing platform specific things.

In this case, Arm64 will treat any "out of bounds" access as "select the value 0", where-as x86/x64 will only do that if the most-significant bit of the index is set. For values where the most-significant bit is not set, it masks the index to be between 0 and Count.

stephentoub replied:
Thanks. Can we include those details in Remarks on the method?

MihaZupan replied:
Something like this?

> Unlike Shuffle, this method does not ensure that values for out-of-range indices are set to zero. Behavior for such indices is platform-dependent.

It feels like it's just repeating the description again, is there something else you had in mind that should be emphasized?

stephentoub replied:
"In this case, Arm64 will treat any "out of bounds" access as "select the value 0", where-as x86/x64 will only do that if the most-significant bit of the index is set. For values where the most-significant bit is not set, it masks the index to be between 0 and Count."

MihaZupan committed:
PR feedback

stephentoub commented:
Thanks!
